src/platform/tiles.c

```diff
@@ -0,0 +1,714 @@
+                        double bestResult = 1e20;
+                        int8_t bestShift = 0;
+                        int8_t midShift = (idx == 2 ? (shifts[0] + shifts[1]) / 2 : 0);
+                        for (int8_t shift = midShift - 5; shift <= midShift + 5; shift++) {
+                            shifts[idx] = shift;
+                            if (processing == 't' || processing == '#') {
+                                shifts[2] = (shifts[0] + shifts[1]) / 2;
+                            }
+                            double blur = prepareTile(surface, baseTileWidth, baseTileHeight, row, column, true);
+                            if (blur < bestResult) {
+                                bestResult = blur;
+                                bestShift = shift;
+                            }
+                        }
+                        shifts[idx] = bestShift;
+                        if (processing == 't' || processing == '#') {
+                            shifts[2] = (shifts[0] + shifts[1]) / 2;
+                        }
+                    }
+                }
+
+                SDL_FreeSurface(surface);
+            }
+
+            // vertical shifts
+            baseTileWidth = MAX_TILE_SIZE;
+            for (baseTileHeight = 7; baseTileHeight <= MAX_TILE_SIZE; baseTileHeight++) {
+                int8_t *shifts = tileShifts[row][column][1][baseTileHeight - 1];
+                SDL_Surface *surface = SDL_CreateRGBSurfaceWithFormat(0, baseTileWidth * TILE_COLS, baseTileHeight * TILE_ROWS, 32, SDL_PIXELFORMAT_ARGB8888);
+                if (!surface) sdlfatal(__FILE__, __LINE__);
+
+                for (int i = 0; i < 3; i++) {
+                    for (int idx = 0; idx < (processing == 't' ? 1 : 3); idx++) {
+                        double bestResult = 1e20;
+                        int8_t bestShift = 0;
+                        int8_t midShift = (idx == 2 ? (shifts[0] + shifts[1]) / 2 : 0);
+                        for (int8_t shift = midShift - 5; shift <= midShift + 5; shift++) {
+                            shifts[idx] = shift;
+                            double blur = prepareTile(surface, baseTileWidth, baseTileHeight, row, column, true);
+                            if (blur < bestResult) {
+                                bestResult = blur;
+                                bestShift = shift;
+                            }
+                        }
+                        shifts[idx] = bestShift;
+                    }
+                }
+
+                SDL_FreeSurface(surface);
+            }
+        }
+    }
+    SDL_SetWindowTitle(Win, "Brogue");
+}
+
+
+static void init() {
+
+    // load the large PNG
+    char filename[BROGUE_FILENAME_MAX];
+    sprintf(filename, "%s/assets/tiles.png", dataDirectory);
+    SDL_Surface *image = IMG_Load(filename);
+    if (!image) imgfatal(__FILE__, __LINE__);
+    TilesPNG = SDL_ConvertSurfaceFormat(image, SDL_PIXELFORMAT_ARGB8888, 0);
```

Comment:
Do you have it backwards? `ARGB888` is a *packed component order*, which means the alpha component is found in the most significant bits of a *32-bit integer*. That's consistent with the rest of my code, which treats pixels as 32-bit integers, with alpha at the top (e.g `*pixel++ = (alpha << 24) | 0xffffffU;` on line 314).

`ARGB32` is an *array order*, which treats a pixel as 4 bytes, with alpha at the lowest address. If I switched to that format, I would have to update my code to change the assumption, from "alpha in the top bits" to "alpha in the first byte".

Array formats (`RGBA32` and the likes) are only useful for serializing pixel data (e.g to disk or network), so that the components come in a predictable order regardless of machine endianness. That's why they weren't part of SDL2 until recently (2.0.5), whereas the former have been around since the beginning. For everything else, packed formats are perfectly suitable and in fact what SDL uses internally (and I believe video drivers too).

I'd be happy to switch to an array format, and modify my code accordingly, *if* someone can prove that `SDL_PIXELFORMAT_ARGB8888` performs worse than `SDL_PIXELFORMAT_BGRA8888` on big-endian machines.

Reply:
Thanks for the thorough explanation - yes I did get the formats the wrong way around, my bad!